File: backend/onGo-domain/src/main/kotlin/com/ongo/domain/subscription/Subscription.kt
package com.ongo.domain.subscription

import com.ongo.common.enums.BillingCycle
import com.ongo.common.enums.PlanType
import com.ongo.common.enums.SubscriptionStatus
import java.time.LocalDateTime

data class Subscription(
    val id: Long? = null,
    val userId: Long,
    val planType: PlanType,
    val status: SubscriptionStatus = SubscriptionStatus.FREE,
    val price: Int = 0,
    val billingCycle: BillingCycle = BillingCycle.MONTHLY,
    val currentPeriodStart: LocalDateTime? = null,
    val currentPeriodEnd: LocalDateTime? = null,
    val nextBillingDate: LocalDateTime? = null,
    val pendingPlanType: PlanType? = null,
    /** 기간 종료 후 함께 적용할 결제 주기. 다운그레이드와 주기 변경을 한 번에 예약할 때만 값이 있다. */
    val pendingBillingCycle: BillingCycle? = null,
    /**
     * PortOne 빌링키(AES-256 암호화). 없으면 정기 청구를 시도할 수 없다.
     *
     * 응답 DTO·로그에 절대 싣지 않는다 — 이 값 하나로 고객에게 반복 청구가 가능해
     * 유출 시 피해가 액세스 토큰보다 크다.
     */
    val billingKeyEncrypted: String? = null,
    val storageQuotaLimitBytes: Long? = null,
    val paddleSubscriptionId: String? = null,
    val paddleCustomerId: String? = null,
    val cancelledAt: LocalDateTime? = null,
    val trialStart: LocalDateTime? = null,
    val trialEnd: LocalDateTime? = null,
    val trialPlanType: PlanType? = null,
    val pausedAt: LocalDateTime? = null,
    val resumeAt: LocalDateTime? = null,
    val createdAt: LocalDateTime? = null,
    val updatedAt: LocalDateTime? = null,
) {
    /**
     * 이 구독을 ACTIVE 로 두기 전에 **채워져 있어야 하는** 청구창 값의 이름.
     * 빈 리스트면 그대로 활성화해도 된다.
     *
     * ## 왜 필요한가
     *
     * 기간이 비어 있는 ACTIVE 유료 구독은 **청구되지도 만료되지도 않는다.** SQL 의 NULL
     * 비교는 참이 아니라 UNKNOWN 이라 `findDueForBilling`(`next_billing_date <= now`) 과
     * `findTrialExpired`(`trial_end <= now`) 가 그 행을 영원히 건너뛴다. 그러면 결제 없이
     * 유료 권한이 무기한 유지되고, 그 손해는 되돌릴 수 없다.
     *
     * 그런 구독이 운영에서 실제로 발견됐다(BUSINESS, 기간 전부 NULL, 결제 0건). 상태를
     * ACTIVE 로 되돌리는 경로가 여럿이라(관리자 활성화, 일시정지 자동 재개) 판단을 한
     * 곳에만 두고 각 경로가 이것을 부른다 — 규칙이 흩어지면 한쪽만 고치게 된다.
     *
     * ## FREE 는 대상이 아니다
     *
     * 무료 구독에는 청구창이 없다. 정상적으로 만들어진 FREE 행부터 기간이 NULL 이므로
     * 여기에 걸면 무료 사용자의 상태 전환이 통째로 막힌다.
     */
    fun missingPaidBillingWindow(): List<String> {
        if (planType == PlanType.FREE) return emptyList()
        return buildList {
            if (currentPeriodStart == null) add("current_period_start")
            if (currentPeriodEnd == null) add("current_period_end")
            if (nextBillingDate == null) add("next_billing_date")
        }
    }
}

File: frontend/src/types/subscription.ts
export type PlanType = 'FREE' | 'STARTER' | 'PRO' | 'BUSINESS'

export type SubscriptionStatus = 'ACTIVE' | 'CANCELLED' | 'PAST_DUE' | 'FREE' | 'TRIALING' | 'PAUSED'

export interface PlanFeatures {
  maxPlatforms: number
  monthlyUploads: number
  scheduleDays: number
  analyticsDays: number
  storageGB: number
  freeCredits: number
  maxTeamMembers: number
}

export interface Subscription {
  id?: number
  userId?: number
  planType: PlanType
  status: SubscriptionStatus
  price: number
  billingCycle: 'MONTHLY' | 'YEARLY'
  currentPeriodEnd: string | null
  nextBillingDate: string | null
  features: PlanFeatures | string[]
  paddleSubscriptionId?: string | null
  paddleCustomerId?: string | null
  trialEnd?: string | null
  pausedAt?: string | null
  resumeAt?: string | null
  pendingPlanType?: PlanType | null
  pendingBillingCycle?: 'MONTHLY' | 'YEARLY' | null
}

export interface ChangePlanResponse {
  subscription: Subscription
  proratedAmount: number | null
  effectiveDate: string
}

export interface Plan {
  type: PlanType
  name: string
  price: number
  yearlyPrice: number
  maxPlatforms: number
  maxUploadsPerMonth: number
  maxScheduleDays: number
  analyticsPeriodDays: number
  storageMb: number
  commentManagement: boolean
  teamMembers: number
  freeAiCredits: number
  support: string
}

/**
 * 로그인 전 가격 안내 전용 상수.
 *
 * **로그인 후 화면에서는 쓰지 않는다.** 가격과 한도는 서버가 결제 기준으로 삼는 값이고,
 * 인증된 `GET /subscriptions/plans` 가 그 값을 내려준다. 화면이 상수를 대신 그리면
 * 한쪽이 뒤처졌을 때 사용자가 본 금액과 청구액이 갈린다.
 *
 * `LoginView` 만 예외다 — 인증 전이라 그 API 를 부를 수 없다. 그 화면의 숫자가 서버와
 * 어긋나지 않도록 `planPricingContract.test.ts` 가 두 값을 맞춰 본다.
 */
export const PLANS: Plan[] = [
  {
    type: 'FREE',
    name: 'Free',
    price: 0,
    yearlyPrice: 0,
    maxPlatforms: 1,
    maxUploadsPerMonth: 5,
    maxScheduleDays: 0,
    analyticsPeriodDays: 7,
    storageMb: 1024,
    commentManagement: false,
    teamMembers: 0,
    freeAiCredits: 30,
    support: '커뮤니티',
  },
  {
    type: 'STARTER',
    name: 'Starter',
    price: 9900,
    yearlyPrice: 99000,
    maxPlatforms: 3,
    maxUploadsPerMonth: 30,
    maxScheduleDays: 7,
    analyticsPeriodDays: 30,
    storageMb: 10240,
    commentManagement: false,
    teamMembers: 0,
    freeAiCredits: 100,
    support: '이메일',
  },
  {
    type: 'PRO',
    name: 'Pro',
    price: 19900,
    yearlyPrice: 199000,
    maxPlatforms: 4,
    maxUploadsPerMonth: 100,
    maxScheduleDays: 30,
    analyticsPeriodDays: 365,
    storageMb: 51200,
    commentManagement: true,
    teamMembers: 2,
    freeAiCredits: 300,
    support: '우선 이메일',
  },
  {
    type: 'BUSINESS',
    name: 'Business',
    price: 49900,
    yearlyPrice: 499000,
    maxPlatforms: 4,
    maxUploadsPerMonth: -1,
    maxScheduleDays: 90,
    analyticsPeriodDays: -1,
    storageMb: 204800,
    commentManagement: true,
    teamMembers: 10,
    freeAiCredits: 1000,
    support: '전담 매니저',
  },
]

export interface Payment {
  id: number
  userId: number
  type: 'SUBSCRIPTION' | 'CREDIT'
  amount: number
  description: string
  /**
   * 서버 `PaymentStatus` 를 그대로 옮긴다.
   *
   * PENDING 이 빠져 있었는데, 체크아웃은 결제창을 열기 **전에** PENDING 행을 만들므로
   * 사용자가 결제를 취소하거나 브라우저를 닫으면 그 상태로 남아 목록에 그대로 나온다.
   * 타입에 없다고 값이 안 오는 것은 아니다.
   */
  status: 'PENDING' | 'COMPLETED' | 'FAILED' | 'REFUNDED'
  /**
   * 서버가 보내는 필드명은 `createdAt` 이다(`PaymentItem`).
   *
   * 예전에는 `paidAt` 으로 읽어서 항상 undefined 였고, 화면의 날짜 칸은 모든 행에서
   * "Invalid Date" 였다. 결제 시각이 아니라 **결제 행이 만들어진 시각**이라는 점도
   * 이름 그대로 두는 편이 정확하다 — PENDING 행에는 결제 시각이 아직 없다.
   */
  createdAt: string | null
  receiptUrl: string | null
}

export interface ChangePlanRequest {
  targetPlan: PlanType
  billingCycle?: 'MONTHLY' | 'YEARLY'
}

export interface PlanInfo {
  planType: PlanType
  price: number
  yearlyPrice: number
  features: PlanFeatures
  recommended: boolean
}

export interface Coupon {
  id: number
  code: string
  description: string | null
  discountType: string
  discountValue: number
  applicablePlans: string | null
  maxUses: number | null
  usedCount: number
  active: boolean
  validFrom: string
  validUntil: string | null
}

/*
 * CouponValidation 은 제거했다. 고객 쿠폰 엔드포인트가 항상 실패하므로 이 응답을 받는
 * 클라이언트가 없다. 남겨두면 "할인 응답 계약이 이미 있다"고 읽힌다.
 */

export interface UsageAlertConfig {
  id: number
  alertType: string
  thresholdPercent: number
  enabled: boolean
  lastAlertedAt: string | null
}
